File: leetcode/reverseLinkedList/guanduo/reverseLinkedList.py
"""
Reverse a singly linked list.

Example:

Input: 1->2->3->4->5->NULL
Output: 5->4->3->2->1->NULL
"""

# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None


"""
Same Idea as iteration. 
Each recursion call extracts the next node, then pass to the recursive call.
If reaching None, return previsou node
"""
class Solution(object):
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        def reverse(node,prev):
            if not node: return prev             #reaching the end, return the last node, which will be the new head
            next_node,node.next = node.next,prev #extract the original next node, continue passing 
            return reverse(next_node,node)
        return reverse(head,None)

File: leetcode/reverseLinkedList/guanduo/reverseLinkedList.cpp
/*
Reverse a singly linked list.

Example:

Input: 1->2->3->4->5->NULL
Output: 5->4->3->2->1->NULL
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

/*
Using Iteration to reverse a linked list.
See python version for the recursive way.
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
      ListNode* prev = NULL; //stores the previous node, since the current node's next will be assigned to the previous one
      ListNode* after;      //store the original next node, since we still iterate through the LL in the original direction

      while(head) {         //while there are still more nodes
	after = head->next; //store the original next node first
	head->next = prev;  //reverse the current node direction
	prev = head;        //store current node as previous node for the next node's reverse operation
	head = after;       //move the current node to the next node (in the original direction)
      }
      return prev;          //since head is NULL now.
    }
};
